src/movies.js

```diff
@@ -1,16 +1,136 @@
+    return directorsClean;
+}
+
+function directorDuplicado(director, indice, array){ //Callback para comprobar si el director se encuentra ya en los elementos previos al índice actual
+for(let i=0; i<indice;i++){
+  if(array[i]===director)return false;
+}
+return true;
+}
+
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
 
+function howManyMovies(movies) {
+    const arrayOfDirectors = movies.filter((objeto)=>(objeto.director==="Steven Spielberg" && objeto.genre.includes("Drama")));
+    return arrayOfDirectors.length;
+  }
+
 // Iteration 3: All rates average - Get the average of all rates with 2 decimals
 
+function ratesAverage(movies){
+    let average = movies.reduce((acc, next)=>{
+        return next.rate? acc + next.rate : acc + 0; //sumamos 0 si no existe puntuación
+
+    },0);
+    average = (average/movies.length) || 0; // media es igual a 0 si no hay películas
+    return Number.parseFloat(average.toFixed(2));
+}
+
 // Iteration 4: Drama movies - Get the average of Drama Movies
 
+function dramaMoviesRate(movies){
+    let dramaMovies = movies.filter((value,index,array)=> value.genre.includes("Drama"));
+    return ratesAverage(dramaMovies);
+}
+
 // Iteration 5: Ordering by year - Order by year, ascending (in growing order)
 
+function orderByYear(movies){
+    let nuevaArray = movies.slice(); //copia del array
+    return nuevaArray.sort((a,b) =>{
+        return a.year===b.year ? a.title.localeCompare(b.title) : a.year-b.year;
+    });
+}
+
 // Iteration 6: Alphabetic Order - Order by title and print the first 20 titles
 
+function orderAlphabetically(movies){
+    let nuevaArray = movies.slice();
+    nuevaArray.sort((a,b) =>{ //ordenamos alfabéticamente los objetos en función de title
+        return a.title.localeCompare(b.title);
+    });
+    return nuevaArray.slice(0,20).map(value => value.title); //cortamos los 20 primeros elementos y mapeamos por title
+    
+}
+
 // BONUS - Iteration 7: Time Format - Turn duration of the movies from hours to minutes
 
+function conversion(horas){ //función auxiliar para hacer la conversión
+    let indiceH = horas.indexOf("h"); //calculamos posición de la h en el formato de la duración
+    let indiceM = horas.indexOf("m"); //calculamos posición de la m en el formato de la duración
+
+    let h;
+    let m;
+
+    if(indiceH===-1){//en el caso de que solo nos pasen minutos
+        m = horas.slice(0,indiceM);
+        return  Number(m);
+    }else{
+        let h = horas.slice(0,indiceH);
+        let m = horas.slice(indiceH+1,indiceM);
+    
+        return Number(h)*60 + Number(m);
+    }
+}
+
+function turnHoursToMinutes(movies){
+    let nuevaArray = movies.slice(); //copia del array
+
+    return nuevaArray.map((objeto)=>{
+        objeto.duration = conversion(objeto.duration); //utilizamos la función auxiliar de conversión en el mapeo
+        return objeto;
+    });
+}
```

Comment:
Molt bè 👍 
El test "Should return a new array, not update the original one" no pasa, porque se ha hecho una "shallow copy" del array, pero los objetos dentro son los mismos.
Para devolver una "deep copy" del array, también se tiene que devolver nuevos objetos. Puedes usar el método Object.assign() como sigue:

```
return nuevaArray.map((objeto)=>{
    const newObjeto = Object.assign({}, objeto);
    newObjeto.duration = conversion(newObjeto.duration); //utilizamos la función auxiliar de conversión en el mapeo
      
      return newObjeto
  });
```